compiler/rustc_lint/src/default_could_be_derived.rs

```diff
@@ -0,0 +1,422 @@
+    ///     x: i32 = 101,
+    ///     y: NonDefault,
+    /// }
+    ///
+    /// struct NonDefault;
+    ///
+    /// #[deny(default_overrides_default_fields)]
+    /// impl Default for Foo {
+    ///     fn default() -> Foo {
+    ///         Foo { x: 100, y: NonDefault }
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Manually writing a `Default` implementation for a type that has
+    /// default field values runs the risk of diverging behavior between
+    /// `Type { .. }` and `<Type as Default>::default()`, which would be a
+    /// foot-gun for users of that type that would expect these to be
+    /// equivalent. If `Default` can't be derived due to some fields not
+    /// having a `Default` implementation, we encourage the use of `..` for
+    /// the fields that do have a default field value.
+    pub DEFAULT_OVERRIDES_DEFAULT_FIELDS,
+    Warn,
+    "detect `Default` impl that should use the type's default field values"
+}
+
+declare_lint_pass!(DefaultCouldBeDerived => [DEFAULT_COULD_BE_DERIVED, DEFAULT_OVERRIDES_DEFAULT_FIELDS]);
+
+impl<'tcx> LateLintPass<'tcx> for DefaultCouldBeDerived {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx hir::Item<'tcx>) {
+        if !cx.tcx.features().default_field_values() {
+            // FIXME: this logic can be extended to check for types without default field values
+            // doing the same as `clippy::derivable_impls`, with slightly more information thanks
+            // to the added "default equivalence" table.
+            return;
+        }
+        let hir = cx.tcx.hir();
+        let hir::ItemKind::Impl(data) = item.kind else { return };
+        let Some(trait_ref) = data.of_trait else { return };
+        let Res::Def(DefKind::Trait, def_id) = trait_ref.path.res else { return };
+        let Some(default_def_id) = cx.tcx.get_diagnostic_item(sym::Default) else { return };
+        if Some(def_id) != Some(default_def_id) {
+            return;
+        }
+        if cx.tcx.has_attr(def_id, sym::automatically_derived) {
+            return;
+        }
+        let hir_self_ty = data.self_ty;
+        let hir::TyKind::Path(hir::QPath::Resolved(_, path)) = hir_self_ty.kind else { return };
+        let Res::Def(def_kind, type_def_id) = path.res else { return };
+
+        let generics = cx.tcx.generics_of(type_def_id);
+        if !generics.own_params.is_empty() && def_kind != DefKind::Enum {
+            // For enums, `#[derive(Default)]` forces you to select a unit variant to avoid
+            // "imperfect derives", unnecessary bounds on type parameters, so even if the enum has
+            // type parameters we can still lint on the manual impl if the return is a unit
+            // variant.
+            return;
+        }
+
+        // We have a manual `impl Default for Ty {}` item, where `Ty` has no type parameters or is
+        // an enum.
+
+        for assoc in data.items {
```

Comment:
everything nested here should be pulled into a function. it's indentation runaway like crazy lol.